### PR TITLE
[v13] Bump github.com/sigstore/cosign/v2 to 2.2.1 in /integrations/kube-agent-updater

### PR DESCRIPTION
**Pull request by jentfoo**
v13 Backport of #34428

### CODE REVIEW

File: integrations/kube-agent-updater/go.mod [hunk 1]
```diff
@@ -4,18 +4,18 @@ go 1.19
...
+	github.com/jonboulle/clockwork v0.4.0
 	github.com/opencontainers/go-digest v1.0.0
-	github.com/sigstore/cosign/v2 v2.0.1-0.20230228132138-2f1ec646de33
-	github.com/sigstore/sigstore v1.6.4
-	github.com/stretchr/testify v1.8.2
-	golang.org/x/exp v0.0.0-20230321023759-10a507213a29
-	golang.org/x/mod v0.10.0
-	k8s.io/api v0.26.1
-	k8s.io/apimachinery v0.26.1
+	github.com/sigstore/cosign/v2 v2.2.1
```
Comment (codingllama) [inline integrations/kube-agent-updater/go.mod:12]: Note that this is not a 2 minors jump.

**Review by codingllama — APPROVED** [review]

**Comment by hugoShaka** [comment]
This PR updates a lot of dependencies, are we really forced to update the kube libraries versions? I'm asking because this could break compatibility with old Kubernetes versions and I'd rather not bump the kube minor version if we can avoid it.

**Comment by jentfoo** [comment]
We do need this update in order to address CVE-2023-46737.  Unfortunately cosign only released the fix as part of 2.2.1, so I don't see any older versions we can patch to.

**Review by hugoShaka — APPROVED** [review]

**Comment by jentfoo** [comment]
@codingllama and @hugoShaka can you help me with this update?  Do the failures here look familiar at all?  I have looked for what change may be missing, but it's not obvious to me.  Thank you!

**Comment by hugoShaka** [comment]
> can you help me with this update? Do the failures here look familiar at all? I have looked for what change may be missing, but it's not obvious to me. v14 has been merged, so comparing the differences there may be useful. Thank you!

According to the linter logs
```
could not import sigs.k8s.io/controller-runtime/pkg/cache (-:
# sigs.k8s.io/controller-runtime/pkg/cache
  /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.4/pkg/cache/multi_namespace_cache.go:308:9:
cannot use handles (variable of type map[string]"k8s.io/client-go/tools/cache".ResourceEventandlerRegistration)
as "k8s.io/client-go/tools/cache".ResourceEventHandlerRegistration value
in return statement: map[string]"k8s.io/client-go/tools/cache".ResourceEventHandlerRegistration
does not implement "k8s.io/client-go/tools/cache".ResourceEventHandlerRegistration (missing method HasSynced)
```

This suggests two incompatible libraries are used together: one does not properly implement the interface used by the other (`missing method HasSynced`). Based on the logs I suspect the incompatibility happens between `sigs.k8s.io/controller-runtime/pkg/cache` and `k8s.io/client-go/tools/cache`.

I'm not a go dependency expert so I cannot advise you on the magic commands to run to fix the broken dependencies. Maybe @codingllama will know a cool trick to fix this.

**Comment by codingllama** [comment]
A quick build attempt tells me the module needs tidying. After `go mod tidy`, it looks like controller-runtime@v0.14.x doesn't build with the new dependencies:

```shell
$ go build ./...
 
# sigs.k8s.io/controller-runtime/pkg/cache
/Users/alan/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.4/pkg/cache/multi_namespace_cache.go:308:9: cannot use handles (variable of type map[string]"k8s.io/client-go/tools/cache".ResourceEventHandlerRegistration) as "k8s.io/client-go/tools/cache".ResourceEventHandlerRegistration value in return statement: map[string]"k8s.io/client-go/tools/cache".ResourceEventHandlerRegistration does not implement "k8s.io/client-go/tools/cache".ResourceEventHandlerRegistration (missing method HasSynced)
/Users/alan/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.4/pkg/cache/multi_namespace_cache.go:321:9: cannot use handles (variable of type map[string]"k8s.io/client-go/tools/cache".ResourceEventHandlerRegistration) as "k8s.io/client-go/tools/cache".ResourceEventHandlerRegistration value in return statement: map[string]"k8s.io/client-go/tools/cache".ResourceEventHandlerRegistration does not implement "k8s.io/client-go/tools/cache".ResourceEventHandlerRegistration (missing method HasSynced)
/Users/alan/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.4/pkg/cache/multi_namespace_cache.go:326:17: impossible type assertion: h.(map[string]toolscache.ResourceEventHandlerRegistration)
	map[string]"k8s.io/client-go/tools/cache".ResourceEventHandlerRegistration does not implement "k8s.io/client-go/tools/cache".ResourceEventHandlerRegistration (missing method HasSynced)
```

Updating controller-runtime itself brings a whole new set of breakages. k8s modules can be pretty finicky to update. I can't say right now what would work, this would take me a while to fix.

**Comment by codingllama** [comment]
A quick blame shows me #31712, which we may need to backport along with this one.

**Comment by jentfoo** [comment]
@codingllama I looked at that PR, however I don't believe it's necessary since the controller is not being updated as part of this.  However I did not test it.

**Comment by codingllama** [comment]
I don't think you'll get this to build without updating the controller.

**Comment by jentfoo** [comment]
I added the changes from https://github.com/gravitational/teleport/pull/27198 and https://github.com/gravitational/teleport/pull/31712.  Both are necessary to update the controller-runtime.

File: integrations/kube-agent-updater/go.mod [hunk 2]
```diff
@@ -1,22 +1,24 @@
 module github.com/gravitational/teleport/integrations/kube-agent-updater
 
-go 1.19
+go 1.21
+
+toolchain go1.21.1
```
Comment (codingllama) [inline integrations/kube-agent-updater/go.mod:5]: v13 is still running on Go 1.20, this won't work.

Comment (jentfoo) [reply integrations/kube-agent-updater/go.mod]: Yes, this was accidental, fixed now

**Comment by jentfoo** [comment]
@hugoShaka With the runtime update this build appears to be passing.  Can you re-review this PR and let me know if you feel there are significant compatibility concerns that we need to consider?

**Comment by hugoShaka** [comment]
Looks good to me

**Review by codingllama — APPROVED** [review]

**Pull request merged**